File: Topological_Sorting/14567_prerequisite.py
# 14567 선수과목
import sys
from collections import deque
input = sys.stdin.readline


def main():
    N, M = map(int, input().split())
    visited = [0] * (N+1)
    prerequisite = [set() for _ in range(N+1)]
    for _ in range(M):
        A, B = map(int, input().split())
        prerequisite[A].add(B)
        visited[B] += 1
    
    ans = [0] * N
    que = deque()
    for i in range(1, N+1):
        if not visited[i]:
            que.append((i, 1))
            ans[i-1] = 1
    
    while que:
        spot, deg = que.popleft()
        for node in tuple(prerequisite[spot]):
            visited[node] -= 1
            if not visited[node]:
                que.append((node, deg+1))
                ans[node-1] = deg+1

    print(*ans)
    return

if __name__ == "__main__":
    main()

File: Topological_Sorting/14567_prerequisite.js
const fs = require("fs");

const setting = () => {
    // const input = fs.readFileSync("input.txt").toString().split("\n");
    const input = fs.readFileSync('/dev/stdin', "utf8").trim().split("\n");
    const [N, M] = input[0].split(" ").map(Number);
    const nearList = [];
    for (let i =0;i<N;i++) nearList.push([]);
    const visited = Array(N).fill(0);
    if (M === 0) return [N , M ,nearList, visited];

    let [a,b] = [0,0];

    input.slice(1, ).forEach((element) => {
        [a, b] = element.split(" ").map(Number);
        nearList[a-1].push(b-1);
        visited[b-1] ++;
    })
    return [N, M, nearList, visited];
}


const main = (N, M, nearList, visited) => {
    const queue = {};
    let front = 0;
    let rear = 0;
    const ans = Array(N).fill(1);

    for (let i=0;i<N;i++){
        if(visited[i] === 0){
            queue[rear++] = [i, 1];
        }
    }

    
    let spot = 0;
    let deg = 1;
    while (queue[front]){
        [spot, deg] = queue[front++];
        delete queue[front-1];
        if (!nearList[spot]) continue;

        for (node of nearList[spot]) {
            visited[node] --;
    
            if (visited[node] === 0) {
                queue[rear++] = [node, deg+1];
                ans[node] = deg+1;
            }
        }
    }   

    console.log(...ans);
}


const [ N, M, nearList, visited ] = setting();
main(N,M,nearList,visited);
